perf: optimize get_num_new_matched_tokens latency in vLLM scheduler

xiaguan opened the pull request:
## Summary

This PR optimizes the latency of the `get_num_new_matched_tokens` function in the vLLM scheduler, stabilizing overall latency to around 1-2ms.

It checks the keys using the `batchexist` RPC interface from the `Mooncake master` (which handles global metadata in Mooncake). This way, it can get hit info from the whole cluster, and the latency tends to be more stable too

## Configuration Requirements

**Important**: LMCache must be configured to use **Mooncake-only storage** for this optimization to work correctly. Using other storage backends will result in inaccurate lookup results.

### Required Configuration Example

```yaml
chunk_size: 256
local_device: "cpu"
remote_url: "mooncakestore://127.0.0.1:50051/"
remote_serde: "naive"
pipelined_backend: False
local_cpu: False
max_local_cpu_size: 5
external_lookup_client: "mooncakestore://127.0.0.1:50051"
```

Shaoting-Feng commented:
Thanks for your PR. Could you provide the performance comparison beween two look up server?

xiaguan commented:
> Thanks for your PR. Could you provide the performance comparison beween two look up server?

Sure, details in this issue https://github.com/LMCache/LMCache/issues/891.

maobaolong commented:
@chunxiaozheng Would you like to take a look at this PR?

Shaoting-Feng reviewed (approved):
LGTM

xiaguan committed:
feat: support mooncake lookup client

Signed-off-by: xiaguan <751080330@qq.com>
Signed-off-by: Jinyang Su <751080330@qq.com>

xiaguan committed:
perf: optimize get_num_new_matched_tokens latency in vLLM scheduler

- Add timing instrumentation to get_num_new_matched_tokens function
- Stabilizes scheduler latency to 1-2ms for improved performance
- Fix config reference bug (vllm_config -> config)
- Requires Mooncake-only storage configuration for accurate results

Signed-off-by: xiaguan <751080330@qq.com>
Signed-off-by: Jinyang Su <751080330@qq.com>

xiaguan committed:
Fix pre-commit formatting issues

- Fix line length violations in utils.py and mooncake_lookup_client.py
- Reformat function signatures to comply with line length limits
- Fix docstring formatting and trailing whitespace
- Add proper newlines at end of files

Signed-off-by: xiaguan <751080330@qq.com>
Signed-off-by: Jinyang Su <751080330@qq.com>

xiaguan committed:
refactor: abstract lookup client creation using factory pattern

Signed-off-by: Jinyang Su <751080330@qq.com>

xiaguan commented:
Conflicts resolved. Thanks for your review 👍

xiaguan commented:
Tested in stable vllm 0.9.2 both LMCacheLookupClient and MooncakeLookupClient.

maobaolong commented on lmcache/v1/config.py:
```diff
@@ -157,6 +161,7 @@ def from_defaults(
         extra_config: Optional[dict] = None,
         save_unfull_chunk: bool = True,
         blocking_timeout_secs: int = 10,
+        mooncake_lookup_client: Optional[str] = None,
```
Can you naming this more abstract so that other backend can implement other `xxx_lookup_client` and can reuse this config?

maobaolong commented on lmcache/v1/lookup_client/factory.py:
```diff
@@ -0,0 +1,109 @@
+        config = lmcache_get_config()
+
+        # Check if mooncake_lookup_client is configured
+        if config.mooncake_lookup_client is not None:
```
If this is `config.external_lookup_client`, it could be designed to be used by variety of implementations, for example, we can use `mooncakestore://<MASTER_ADDRESS>` or `infinistore://<INFINISTORE_ADDRESS>` as the content of `external_lookup_client`.

This idea is referent of remote backend framework.

xiaguan replied:
Thanks for the suggestion — it's been fixed!

maobaolong reviewed (commented):
@xiaguan Left a refactor suggestion, otherwise, it look wonderful!

xiaguan committed:
refactor(config): support external lookup client with uri format

Signed-off-by: Jinyang Su <751080330@qq.com>

Shaoting-Feng commented on lmcache/v1/lookup_client/factory.py:
```diff
@@ -0,0 +1,162 @@
+
+# First Party
+from lmcache.integration.vllm.utils import lmcache_get_config
+from LMCache.lmcache.v1.lookup_client.mooncake_lookup_client import MooncakeLookupClient
```
No module named 'LMCache'

xiaguan replied:
Thks. Fixed

xiaguan committed:
fix import issue

Signed-off-by: Jinyang Su <751080330@qq.com>

maobaolong reviewed (approved):
LGTM @xiaguan Thanks for this great work!

maobaolong committed:
Merge branch 'dev' into mooncake_lookup
